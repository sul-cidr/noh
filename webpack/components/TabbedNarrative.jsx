import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { Markup } from "interweave";
import { Tab, Tabs, TabList, TabPanel } from "react-tabs";
import { setNarrativeTab } from "../actionCreators";

class TabbedNarrative extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      chunks: null,
      titles: null
    };

    const hookupMouseovers = function hookupMouseovers(container) {
      container
        .querySelectorAll("[data-highlight-shodans]")
        .forEach(targetElem => {
          targetElem.addEventListener("mouseover", (/* event */) => {
            const sectionIndices = targetElem.attributes[
              "data-highlight-shodans"
            ].value.split(",");
            sectionIndices.forEach(index => {
              container
                .querySelectorAll(`[data-index="${index}"]`)
                .forEach(shodanBlock =>
                  shodanBlock.classList.add("shodan-map__item--highlight")
                );
            });
          });
          targetElem.addEventListener("mouseout", (/* event */) => {
            const sectionIndices = targetElem.attributes[
              "data-highlight-shodans"
            ].value.split(",");
            sectionIndices.forEach(index => {
              container
                .querySelectorAll(`[data-index="${index}"]`)
                .forEach(shodanBlock =>
                  shodanBlock.classList.remove("shodan-map__item--highlight")
                );
            });
          });
        });
    };

    /* This function is triggered on all tab changes, as well as some other
     * events. It is used to scroll the visible tab panel to the top upon
     * activation, if necessary. */
    this.handleDomRef = tabsRef => {
      if (tabsRef !== null) {
        const appContainer = tabsRef.closest(".app-container");
        if (appContainer) hookupMouseovers(appContainer);
        tabsRef
          .querySelectorAll("a[_onclick]")
          .forEach(link =>
            link.setAttribute("onclick", link.getAttribute("_onclick"))
          );
        const { parentElement } = tabsRef;
        parentElement.scrollTop = 0;
      }
    };
  }

  componentWillMount() {
    const [chunks, titles] = this.parseNarrative();
    this.setState({ chunks, titles });
  }

  parseNarrative() {
    const nar = document.createElement("div");
    nar.innerHTML = this.props.narrative;
    const sections = nar.querySelectorAll("section");
    const chunks = Array.from(sections).map(elt => elt.outerHTML);

    const re = /title=['"][\w\s]+['"]/;
    const titles = [];
    for (let i = 0; i < chunks.length; i += 1) {
      const titleString = re.exec(chunks[i]);
      const pieces = titleString[0].split("=");
      const title = pieces[1].slice(1, -1);
      chunks[i] = chunks[i].replace(`title="${title}"`, "");
      titles.push(title);
    }
    return [chunks, titles];
  }

  render() {
    const narrativeTabList = this.state.titles.map(title => (
      <Tab key={title}>{title}</Tab>
    ));
    const narrativeTabs = this.state.chunks.map(chunk => (
      <TabPanel key={chunk}>
        <Markup
          tagName="div"
          allowAttributes
          content={chunk.replace(/\bonclick=/g, "_onclick=")}
        />
      </TabPanel>
    ));
    const { updateNarrativeTab } = this.props;
    let { narrativeTab } = this.props;
    narrativeTab = Math.min(narrativeTab, this.state.titles.length - 1);
    return (
      <Tabs
        domRef={this.handleDomRef}
        selectedIndex={narrativeTab}
        onSelect={index => updateNarrativeTab(index)}
      >
        <TabList>{narrativeTabList}</TabList>
        {narrativeTabs}
      </Tabs>
    );
  }
}

TabbedNarrative.propTypes = {
  narrative: PropTypes.string.isRequired,
  narrativeTab: PropTypes.number.isRequired,
  updateNarrativeTab: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => ({
  updateNarrativeTab: index => dispatch(setNarrativeTab(index))
});

const mapStateToProps = state => ({
  narrativeTab: state.narrativeTab
});

export const Unwrapped = TabbedNarrative;
export default connect(mapStateToProps, mapDispatchToProps)(TabbedNarrative);
