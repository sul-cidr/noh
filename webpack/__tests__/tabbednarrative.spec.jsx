import React from "react";
import { shallow, mount } from "enzyme";
import TabbedNarrative from "../components/TabbedNarrative";

describe("<TabbedNarrative>", () => {
  const tabbedAnalysis =
    "<div title='section one with many words in title' class='tabbed-narrative'><p>Lorem ipsum dolor sit amet.</p></div><br /><div title='section two'><p>Different stuff</p></div>";

  it("renders as expected", () => {
    const component = shallow(<TabbedNarrative narrative={tabbedAnalysis} />);
    expect(component).toMatchSnapshot();
  });

  it("ref handler runs with tabs ref", () => {
    const component = mount(
      <div className="narrative" role="presentation">
        <TabbedNarrative narrative={tabbedAnalysis} />
      </div>
    );
    TabbedNarrative.handleDomRef(
      component.find("TabbedNarrative").getDOMNode()
    );
  });

  it("ref handler runs with null input", () => {
    TabbedNarrative.handleDomRef(null);
  });
});
