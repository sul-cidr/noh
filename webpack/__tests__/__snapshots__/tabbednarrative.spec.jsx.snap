// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<TabbedNarrative> renders as expected 1`] = `
<Tabs
  defaultFocus={false}
  defaultIndex={null}
  domRef={[Function]}
  forceRenderTabPanel={false}
  onSelect={[Function]}
  selectedIndex={0}
>
  <TabList
    className="react-tabs__tab-list"
  >
    <Tab
      className="react-tabs__tab"
      disabledClassName="react-tabs__tab--disabled"
      focus={false}
      id={null}
      key="section one with many words in title"
      panelId={null}
      selected={false}
      selectedClassName="react-tabs__tab--selected"
    >
      section one with many words in title
    </Tab>
    <Tab
      className="react-tabs__tab"
      disabledClassName="react-tabs__tab--disabled"
      focus={false}
      id={null}
      key="section two"
      panelId={null}
      selected={false}
      selectedClassName="react-tabs__tab--selected"
    >
      section two
    </Tab>
  </TabList>
  <TabPanel
    className="react-tabs__tab-panel"
    forceRender={false}
    key="<section  class="tabbed-narrative"><p>Lorem ipsum dolor <a href="#0" onclick="window.open()">sit</a> amet.</p></section>"
    selectedClassName="react-tabs__tab-panel--selected"
  >
    <Markup
      allowAttributes={true}
      content="<section  class="tabbed-narrative"><p>Lorem ipsum dolor <a href="#0" _onclick="window.open()">sit</a> amet.</p></section>"
      emptyContent={null}
      parsedContent={null}
      tagName="div"
    />
  </TabPanel>
  <TabPanel
    className="react-tabs__tab-panel"
    forceRender={false}
    key="<section ><p>Different stuff</p></section>"
    selectedClassName="react-tabs__tab-panel--selected"
  >
    <Markup
      allowAttributes={true}
      content="<section ><p>Different stuff</p></section>"
      emptyContent={null}
      parsedContent={null}
      tagName="div"
    />
  </TabPanel>
</Tabs>
`;
